web/modules/custom/joinup_search/src/Plugin/search_api/processor/JoinupEntityLatestRevision.php

```diff
@@ -0,0 +1,111 @@
+use Drupal\state_machine_revisions\RevisionManagerInterface;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Ensures that the item to be indexed is the latest version.
+ *
+ * @SearchApiProcessor(
+ *   id = "joinup_entity_latest_revision",
+ *   label = @Translation("Joinup entity latest revision"),
+ *   description = @Translation("Ensures that the version of the entity is the latest revision."),
+ *   stages = {
+ *     "alter_items" = -20,
+ *   },
+ * )
+ */
+class JoinupEntityLatestRevision extends ProcessorPluginBase {
+
+  /**
+   * The entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The state machine revisions manager service.
+   *
+   * @var \Drupal\state_machine_revisions\RevisionManagerInterface
+   */
+  protected $revisionManager;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, array $plugin_definition, EntityTypeManagerInterface $entity_type_manager, RevisionManagerInterface $revision_manager) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->entityTypeManager = $entity_type_manager;
+    $this->revisionManager = $revision_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager'),
+      $container->get('state_machine_revisions.revision_manager')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function supportsIndex(IndexInterface $index) {
+    $supported_entity_types = ['node', 'rdf_entity'];
+    foreach ($index->getDatasources() as $datasource) {
+      if (in_array($datasource->getEntityTypeId(), $supported_entity_types)) {
+        return TRUE;
+      }
+    }
+    return FALSE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function alterIndexedItems(array &$items) {
+    /** @var \Drupal\search_api\Item\Item[] $items */
+    foreach ($items as $item_id => $item) {
+      $original_object = $item->getOriginalObject();
+      $object = $original_object->getValue();
+      if ($object instanceof NodeInterface) {
+        if (!$this->revisionManager->isLatestRevision($object)) {
+          $latest = $this->revisionManager->loadLatestRevision($object);
+          if (!empty($latest)) {
+            $original_object->setValue($latest);
+            $item->setOriginalObject($original_object);
+            $items[$item_id] = $item;
+          }
+        }
+      }
+      elseif ($object instanceof RdfInterface) {
+        if ($object->isPublished()) {
+          /** @var \Drupal\rdf_entity\Entity\RdfEntitySparqlStorage $rdf_storage */
+          $rdf_storage = $this->entityTypeManager->getStorage('rdf_entity');
```

Comment:
This is the only instance in this class where the entity type manager is used. If you are referring to the complete 'get rdf version' maybe we should as I struggled as well but I think we can leave it for now. We have created something similar for the delete but I don't want to make this addition right now in the scope of such a big ticket.